File: src/main/java/com/ada/shared/observability/TokenUsageEstimator.java
package com.ada.shared.observability;

import com.ada.conversation.application.dto.*;
import java.nio.charset.StandardCharsets;
import java.util.*;
import org.springframework.stereotype.Component;

@Component
public class TokenUsageEstimator {
  public long estimate(ContextState c) {
    long n = 0;
    for (var m : c.messages()) n += estimateTokens(m.content());
    for (var t : c.tools())
      n += estimateTokens(t.name() + " " + t.description() + " " + t.inputSchema());
    return n;
  }

  public List<TokenUsageComponent> components(LlmRequest r) {
    var map = new LinkedHashMap<String, Long>();
    for (var m : r.messages())
      map.merge(m.component().name().toLowerCase(), estimateTokens(m.content()), Long::sum);
    long tools =
        r.tools().stream()
            .mapToLong(
                t -> estimateTokens(t.name() + " " + t.description() + " " + t.inputSchema()))
            .sum();
    if (tools > 0) map.put("tools", tools);
    var result = new ArrayList<TokenUsageComponent>();
    map.forEach((k, v) -> result.add(new TokenUsageComponent(k, v, TokenUsageSource.ESTIMATED)));
    result.add(
        new TokenUsageComponent(
            "total",
            result.stream().mapToLong(TokenUsageComponent::tokens).sum(),
            TokenUsageSource.ESTIMATED));
    return result;
  }

  private long estimateTokens(String s) {
    return (s.getBytes(StandardCharsets.UTF_8).length + 3L) / 4L;
  }
}

File: src/main/kotlin/com/ada/shared/observability/TokenUsageEstimator.kt
package com.ada.shared.observability

import com.ada.conversation.application.dto.LlmRequest
import com.ada.conversation.application.dto.TokenUsageComponent
import com.ada.conversation.application.dto.TokenUsageSource
import org.springframework.stereotype.Component

@Component
class TokenUsageEstimator {
    fun components(request: LlmRequest): List<TokenUsageComponent> {
        val messageComponents = request.messages.groupingBy { it.component }
            .fold(0L) { total, message -> total + estimateTokens(message.content) }
            .map { (component, tokens) ->
                TokenUsageComponent(component.name.lowercase(), tokens, TokenUsageSource.ESTIMATED)
            }
        val toolTokens = request.tools.sumOf {
            estimateTokens("${it.name} ${it.description} ${it.inputSchema}")
        }
        val components = if (toolTokens == 0L) messageComponents else {
            messageComponents + TokenUsageComponent("tools", toolTokens, TokenUsageSource.ESTIMATED)
        }
        return components + TokenUsageComponent(
            "total",
            components.sumOf { it.tokens },
            TokenUsageSource.ESTIMATED,
        )
    }

    private fun estimateTokens(value: String): Long = (value.toByteArray().size + 3L) / 4L
}
